### ISSUE TITLE
octet_string_set_to_zero does not provide the security its callers seem to expect

### ISSUE THREAD

**Issue by mmdriley** [cisco/libsrtp#157]
`octet_string_set_to_zero` is used ubiquitously to remove keys or other sensitive data from memory, e.g:
- [crypto/cipher/aes_gcm_ossl.c#L158](https://github.com/cisco/libsrtp/blob/ce18c4f9bd4a89cead4829b8b1bcfc3149439b5f/crypto/cipher/aes_gcm_ossl.c#L158)
- [crypto/cipher/aes_icm_ossl.c#L201](https://github.com/cisco/libsrtp/blob/ce18c4f9bd4a89cead4829b8b1bcfc3149439b5f/crypto/cipher/aes_icm_ossl.c#L201)
- [crypto/cipher/aes_icm.c#L169](https://github.com/cisco/libsrtp/blob/ce18c4f9bd4a89cead4829b8b1bcfc3149439b5f/crypto/cipher/aes_icm.c#L169)
- [crypto/hash/hmac_ossl.c#L117](https://github.com/cisco/libsrtp/blob/ce18c4f9bd4a89cead4829b8b1bcfc3149439b5f/crypto/hash/hmac_ossl.c#L117)
- [crypto/hash/hmac.c#L101](https://github.com/cisco/libsrtp/blob/ce18c4f9bd4a89cead4829b8b1bcfc3149439b5f/crypto/hash/hmac.c#L101)
- [srtp/srtp.c](https://github.com/cisco/libsrtp/blob/ce18c4f9bd4a89cead4829b8b1bcfc3149439b5f/srtp/srtp.c#L714) (at least a dozen times in `srtp_stream_init_keys`)

Here's the [implementation](https://github.com/cisco/libsrtp/blob/ce18c4f9bd4a89cead4829b8b1bcfc3149439b5f/crypto/math/datatypes.c#L453-L461):

```
void
octet_string_set_to_zero(uint8_t *s, int len) {
  uint8_t *end = s + len;

  do {
    *s = 0;
  } while (++s < end);

}
```

Here's a [good writeup from CERT-CC](https://www.securecoding.cert.org/confluence/display/c/MSC06-C.+Beware+of+compiler+optimizations) on why this doesn't work. In short: a reasonably aggressive compiler can optimize this entire thing away once it notices the buffer involved won't be accessed again before being freed.

In some cases, `octet_string_set_to_zero` is actually zeroing a buffer _before_ it's used. Those uses can stay, but all others -- including all the ones mentioned above -- should move to something akin to `OPENSSL_cleanse`, `SecureZeroMemory`, `memset_s`, etc.

**Comment by thisisG** [comment]
Hey! There is a PR up now - care to take a look?
There is a more in depth reasoning in the first commit of the series.

**Comment by pabuhler** [comment]
addressed in #239

**Comment by mmdriley** [comment]
@thisisG Thanks for the fix!
